torch/nn/utils/parametrize.py

```diff
@@ -53,72 +55,211 @@ def cached():
-                        "Assigning to a parametrized tensor is only possible when all the parametrizations "
-                        "implement a 'right_inverse' method.".format(module.__class__.__name__)
+                # else we assume that right_inverse is the identity
+            if self.ntensors == 1:
+                # These exceptions should only throw when a right_inverse function does not
+                # return the same dtype for every input, which should most likely be caused by a bug
+                if not isinstance(value, Tensor):
+                    raise ValueError(
+                        f"`right_inverse` should return a tensor. Got {type(value).__name__}"
+                    )
+                if value.dtype != self.original.dtype:
+                    raise ValueError(
+                        f"The tensor returned by `right_inverse` has dtype {value.dtype} "
+                        f"while `original` has dtype {self.original.dtype}"
+                    )
+                # We know that the result is going to have the same dtype
+                self.original.set_(value)  # type: ignore[call-overload]
+            else:
+                if not isinstance(value, collections.Sequence):
+                    raise ValueError(
+                        "'right_inverse' must return a sequence of tensors. "
+                        f"Got {type(value).__name__}."
                     )
-            self.original.copy_(value)
+                if len(value) != self.ntensors:
+                    raise ValueError(
+                        "'right_inverse' must return a sequence of tensors of length "
+                        f"{self.ntensors}. Got a sequence of lenght {len(value)}."
+                    )
+                for i, tensor in enumerate(value):
+                    original_i = getattr(self, f"original{i}")
+                    if not isinstance(tensor, Tensor):
+                        raise ValueError(
+                            f"`right_inverse` must return a sequence of tensors. "
+                            f"Got element {i} of type {type(tensor).__name__}"
+                        )
+                    if original_i.dtype != tensor.dtype:
+                        raise ValueError(
+                            f"Tensor {i} returned by `right_inverse` has dtype {tensor.dtype} "
+                            f"while `original{i}` has dtype {original_i.dtype}"
+                        )
+                    original_i.set_(tensor)
 
     def forward(self) -> Tensor:
-        x = self.original
-        for module in self:
+        # Unpack the originals for the first parametrization
+        if self.ntensors == 1:
+            x = self[0](self.original)
+        else:
+            originals = (getattr(self, f"original{i}") for i in range(self.ntensors))
+            x = self[0](*originals)
+        # It's not possible to call self[1:] here, so we have to be a bit more cryptic
```

Comment:
nit: if you want to avoid using the internal API, you can use iter/next to drop the first element of the iterator. But I'm not sure that would be less cryptic :p

Reply:
Good point, but as you said, I don't think that it'd be much better. I think we can leave as it is.